File: packages/shared/src/schemas/ingestion-event.ts
/**
 * IngestionEvent Schema
 * Represents data ingestion events from various sources
 * 
 * @version 1.0.0
 * @backward-compatibility
 * - v1.0.0: Initial schema
 */

import { z } from "zod";
import {
  createVersionedSchema,
  statusSchema,
  severitySchema,
  CURRENT_SCHEMA_VERSION,
} from "./common.js";

// ============================================
// SCHEMA DEFINITION
// ============================================

export const ingestionSourceTypeSchema = z.enum([
  "nadfun_api",
  "monad_rpc",
  "social_api",
  "news_api",
  "websocket",
  "webhook",
  "scraper",
  "manual",
]);

export type IngestionSourceType = z.infer<typeof ingestionSourceTypeSchema>;

export const ingestionDataTypeSchema = z.enum([
  "token_data",
  "market_data",
  "transaction",
  "block",
  "news",
  "social",
  "price",
  "liquidity",
  "holder_data",
  "contract_event",
]);

export type IngestionDataType = z.infer<typeof ingestionDataTypeSchema>;

export const ingestionEventSchema = createVersionedSchema({
  // Source information
  sourceType: ingestionSourceTypeSchema,
  sourceId: z.string(),
  sourceName: z.string(),
  sourceUrl: z.string().url().optional(),
  
  // Data classification
  dataType: ingestionDataTypeSchema,
  dataSubtype: z.string().optional(),
  
  // Payload
  payload: z.record(z.unknown()),
  payloadSize: z.number().int().min(0),
  payloadHash: z.string().optional(),
  
  // Processing
  status: statusSchema,
  processingStartedAt: z.string().datetime().optional(),
  processingCompletedAt: z.string().datetime().optional(),
  processingDurationMs: z.number().int().min(0).optional(),
  
  // Error handling
  errorMessage: z.string().optional(),
  errorCode: z.string().optional(),
  retryCount: z.number().int().min(0).default(0),
  maxRetries: z.number().int().min(0).default(3),
  
  // Quality metrics
  dataQualityScore: z.number().min(0).max(1).optional(),
  isValid: z.boolean().default(true),
  validationErrors: z.array(z.string()).default([]),
  
  // Priority
  priority: severitySchema.default("medium"),
  
  // Deduplication
  deduplicationKey: z.string().optional(),
  isDuplicate: z.boolean().default(false),
  
  // Batch info
  batchId: z.string().uuid().optional(),
  batchIndex: z.number().int().min(0).optional(),
  
  // Timestamps
  ingestedAt: z.string().datetime(),
  dataTimestamp: z.string().datetime().optional(),
});

export type IngestionEvent = z.infer<typeof ingestionEventSchema>;

// ============================================
// FACTORY FUNCTIONS
// ============================================

export function createIngestionEvent(
  data: Omit<IngestionEvent, "id" | "schemaVersion" | "createdAt">
): IngestionEvent {
  return ingestionEventSchema.parse({
    ...data,
    id: crypto.randomUUID(),
    schemaVersion: CURRENT_SCHEMA_VERSION,
    createdAt: new Date().toISOString(),
  });
}

// ============================================
// EXAMPLES & FIXTURES
// ============================================

export const ingestionEventExamples: IngestionEvent[] = [
  {
    schemaVersion: "1.0.0",
    id: "550e8400-e29b-41d4-a716-446655440020",
    createdAt: "2024-01-15T12:00:00Z",
    sourceType: "nadfun_api",
    sourceId: "nadfun-trending",
    sourceName: "nad.fun Trending API",
    sourceUrl: "https://api.nadapp.net/api/v1/market/trending",
    dataType: "market_data",
    dataSubtype: "trending_tokens",
    payload: {
      tokens: [
        { symbol: "PEPE", price: 0.00001234, volume24h: 125000 },
        { symbol: "DOGE", price: 0.00005678, volume24h: 89000 },
      ],
    },
    payloadSize: 256,
    payloadHash: "0xabcdef1234567890",
    status: "completed",
    processingStartedAt: "2024-01-15T12:00:00Z",
    processingCompletedAt: "2024-01-15T12:00:01Z",
    processingDurationMs: 1000,
    retryCount: 0,
    maxRetries: 3,
    dataQualityScore: 0.98,
    isValid: true,
    validationErrors: [],
    priority: "high",
    isDuplicate: false,
    ingestedAt: "2024-01-15T12:00:00Z",
    dataTimestamp: "2024-01-15T11:59:55Z",
  },
];

File: services/ingestion/src/schemas/ingestion_event.rs
//! IngestionEvent Schema
//! 
//! Represents data ingestion events from various sources
//! Compatible with TypeScript IngestionEvent schema

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use super::common::{Status, Severity, Uuid, Timestamp, SchemaVersion};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum IngestionSourceType {
    NadfunApi,
    MonadRpc,
    SocialApi,
    NewsApi,
    Websocket,
    Webhook,
    Scraper,
    Manual,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum IngestionDataType {
    TokenData,
    MarketData,
    Transaction,
    Block,
    News,
    Social,
    Price,
    Liquidity,
    HolderData,
    ContractEvent,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IngestionEvent {
    // Base fields
    pub schema_version: SchemaVersion,
    pub id: Uuid,
    pub created_at: Timestamp,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<Timestamp>,
    
    // Source information
    pub source_type: IngestionSourceType,
    pub source_id: String,
    pub source_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_url: Option<String>,
    
    // Data classification
    pub data_type: IngestionDataType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data_subtype: Option<String>,
    
    // Payload
    pub payload: HashMap<String, serde_json::Value>,
    pub payload_size: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payload_hash: Option<String>,
    
    // Processing
    pub status: Status,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub processing_started_at: Option<Timestamp>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub processing_completed_at: Option<Timestamp>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub processing_duration_ms: Option<u64>,
    
    // Error handling
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_code: Option<String>,
    #[serde(default)]
    pub retry_count: u32,
    #[serde(default = "default_max_retries")]
    pub max_retries: u32,
    
    // Quality metrics
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data_quality_score: Option<f64>,
    #[serde(default = "default_true")]
    pub is_valid: bool,
    #[serde(default)]
    pub validation_errors: Vec<String>,
    
    // Priority
    #[serde(default = "default_priority")]
    pub priority: Severity,
    
    // Deduplication
    #[serde(skip_serializing_if = "Option::is_none")]
    pub deduplication_key: Option<String>,
    #[serde(default)]
    pub is_duplicate: bool,
    
    // Batch info
    #[serde(skip_serializing_if = "Option::is_none")]
    pub batch_id: Option<Uuid>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub batch_index: Option<u32>,
    
    // Timestamps
    pub ingested_at: Timestamp,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data_timestamp: Option<Timestamp>,
}

fn default_max_retries() -> u32 {
    3
}

fn default_true() -> bool {
    true
}

fn default_priority() -> Severity {
    Severity::Medium
}

impl IngestionEvent {
    pub fn new(
        source_type: IngestionSourceType,
        source_id: String,
        source_name: String,
        data_type: IngestionDataType,
        payload: HashMap<String, serde_json::Value>,
    ) -> Self {
        let payload_json = serde_json::to_string(&payload).unwrap_or_default();
        let payload_size = payload_json.len() as u64;
        let now = chrono::Utc::now().to_rfc3339();
        
        Self {
            schema_version: super::CURRENT_SCHEMA_VERSION.to_string(),
            id: uuid::Uuid::new_v4().to_string(),
            created_at: now.clone(),
            updated_at: None,
            source_type,
            source_id,
            source_name,
            source_url: None,
            data_type,
            data_subtype: None,
            payload,
            payload_size,
            payload_hash: None,
            status: Status::Pending,
            processing_started_at: None,
            processing_completed_at: None,
            processing_duration_ms: None,
            error_message: None,
            error_code: None,
            retry_count: 0,
            max_retries: 3,
            data_quality_score: None,
            is_valid: true,
            validation_errors: vec![],
            priority: Severity::Medium,
            deduplication_key: None,
            is_duplicate: false,
            batch_id: None,
            batch_index: None,
            ingested_at: now,
            data_timestamp: None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ingestion_event_serialization() {
        let mut payload = HashMap::new();
        payload.insert("key".to_string(), serde_json::json!("value"));
        
        let event = IngestionEvent::new(
            IngestionSourceType::NadfunApi,
            "nadfun-trending".to_string(),
            "nad.fun Trending API".to_string(),
            IngestionDataType::MarketData,
            payload,
        );

        let json = serde_json::to_string_pretty(&event).unwrap();
        println!("{}", json);
        
        assert!(json.contains("sourceType"));
        assert!(json.contains("dataType"));
    }
}
